conformalInference/R/jack.R

```diff
@@ -111,56 +135,141 @@ conformal.pred.jack = function(x, y, x0, train.fun, predict.fun, alpha=0.1,
+    
+    mad.trained<-one.lapply(1:n, function(i){ 
+    
+    r = abs(y[-i] - matrix(predict.fun(updated_models[[i]],x[-i,,drop=F]),
+                           nrow=n-1))
+    
+    return(one.lapply(1:m, function(l)  mad.train.fun(x[-i,,drop=F],r[,l])))
+     })
+    
+  }## mad.trained is a list  n x m
+  
+  
   # Compute leave-one-out residuals by brute-force
   else {
-    res = matrix(0,n,m)
-    for (i in 1:n) {
-      if (verbose) {
-        cat(sprintf("\r%sProcessing training point %i (of %i) ...",txt,i,n))
-        flush.console()
-      }
-     
-      xx = x[-i,]; yy = y[-i]
-      out.i = train.fun(xx,yy)
-      res[i,] = abs(y[i] - predict.fun(out.i,x[i,,drop=F]))
+    
+      res=t(one.sapply(1:n,function(i){ 
+        
+        if (verbose) {
+          cat(sprintf("\r%sProcessing training point %i (of %i) ...",txt,i,n))
+          flush.console()
+        }
+        
+        test_val = matrix(x[i,],nrow=1)
+        pred_values = c(predict.fun(updated_models[[i]] ,  test_val ))
+        val=abs(y[i]- pred_values) ## LOO residual
+        
+        # Local scoring?
+        if (!is.null(mad.train.fun) && !is.null(mad.predict.fun)) {
+          for (l in 1:m) {
+            val[i,l] = val[i,l] / mad.predict.fun(mad.trained[[i]][[l]],x[i,,drop=F])
+          }
+        }
+        
+        if (verbose) cat("\n")
+        
+        return(val)
+      }))
+      
+      ##res is matrix n x m
+      
 
-      # Local scoring?
-      if (!is.null(mad.train.fun) && !is.null(mad.predict.fun)) {
-        r = abs(y[-i] - matrix(predict.fun(out.i,x[-i,,drop=F]),nrow=n-1))
-        for (l in 1:m) {
-          out.mad.i = mad.train.fun(x[-i,,drop=F],r[,l])
-          res[i,l] = res[i,l] / mad.predict.fun(out.mad.i,x[i,,drop=F])
+  }
+    
+
+    lo = up = matrix(0,n0,m)
+  
+    
+    if(plus){
+      
+      
+      ## First compute local scaling
+      mad.x0.vec<-one.lapply (1:m, function(l) one.sapply(1:n, function(k){
+        
+        # Local scoring?
+        if (!is.null(mad.train.fun) && !is.null(mad.predict.fun)) {
+          return(mad.predict.fun(mad.trained[[k]][[l]],x0))
         }
-      }
+        else {
+          return(rep(1,n0))
+        }
+      })) ## it is a list of m elements, each containing a matrix n x n0
+      
+      
+      # Compute bounds
+       lo_up = t(one.sapply(1:n0, function(ii){ 
+         
+        #Compute new responses removing one observation from models at each time
+        mu=matrix(one.sapply(1:n, function(jj){
+          return(c(predict.fun(updated_models [[jj]],matrix(x0[ii,],nrow=1))))
+        }),nrow=n)  ## matrix of dim n x m
+        
+        
+        # Get quantiles
+        lo_vec<-one.sapply(1:m, function(l) mu[,l]-res[,l]*mad.x0.vec[[l]][,ii])
+        up_vec<-one.sapply(1:m, function(l) mu[,l]+res[,l]*mad.x0.vec[[l]][,ii])
+        lo.obs <- one.sapply(1:m, function(l) quantile(lo_vec[,l],alpha))
+        up.obs <- one.sapply(1:m, function(l) quantile(up_vec[,l],1-alpha))
```

Comment:
We should be using the proper corrections for finite-samples here as in the jackknife+ paper. For example `ceil((n+1)*(1-alpha))/n` for the upper quantile. And the lower one is similar, I think it is `floor((n+1)*alpha)/n`. But this is from memory: please double-check that paper to be sure.